cpp/src/arrow/util/async_nursery.cc

```diff
@@ -0,0 +1,123 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/util/async_nursery.h"
+
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace util {
+
+AsyncCloseable::AsyncCloseable() : on_closed_(Future<>::Make()) {}
+AsyncCloseable::AsyncCloseable(AsyncCloseable* parent) : on_closed_(Future<>::Make()) {
+  parent->AddDependentTask(OnClosed());
+}
+
+AsyncCloseable::~AsyncCloseable() {
+  DCHECK_NE(nursery_, nullptr) << "An AsyncCloseable must be created with a nursery "
+                                  "using MakeSharedCloseable or MakeUniqueCloseable";
+}
+
+const Future<>& AsyncCloseable::OnClosed() { return on_closed_; }
+
+void AsyncCloseable::AddDependentTask(const Future<>& task) {
+  DCHECK(!closed_);
+  if (num_tasks_outstanding_.fetch_add(1) == 1) {
+    tasks_finished_ = Future<>::Make();
+  }
+  task.AddCallback([this](const Status& st) {
+    if (num_tasks_outstanding_.fetch_sub(1) == 1 && closed_.load()) {
+      tasks_finished_.MarkFinished(st);
+    }
+  });
+}
+
+void AsyncCloseable::SetNursery(Nursery* nursery) { nursery_ = nursery; }
+
+void AsyncCloseable::Destroy() {
+  DCHECK_NE(nursery_, nullptr);
+  closed_ = true;
+  nursery_->num_closeables_destroyed_.fetch_add(1);
+  Future<> finish_fut;
+  if (tasks_finished_.is_valid()) {
+    if (num_tasks_outstanding_.fetch_sub(1) > 1) {
+      finish_fut = AllComplete({DoClose(), tasks_finished_});
+    } else {
+      // Any added tasks have already finished so there is nothing to wait for
+      finish_fut = DoClose();
+    }
+  } else {
+    // No dependent tasks were added
+    finish_fut = DoClose();
+  }
+  finish_fut.AddCallback([this](const Status& st) {
+    if (on_closed_.is_valid()) {
+      on_closed_.MarkFinished(st);
+    }
+    nursery_->OnTaskFinished(st);
+    delete this;
```

Comment:
Ok, so this mandates that this object is heap-allocated using the default C++ allocator, right? Can you mention this somewhere in the docstring?